### ISSUE TITLE
Setting up gateway memory cache correctly

### ISSUE THREAD

**Issue by vstax** [leo-project/leofs#936]
In our cluster, objects are often requested right after PUT operation, and I want objects uploaded through some gateway (including large objects) to be stored for at least some time in RAM, so they can be returned from cache, reducing latency and load on storage. There is external cache (squid) between clients and gateway so I don't need objects to be stored for a long time, even something like 10 seconds helps a lot. Because of this there is also no disk cache on gateway (squid has large disk cache, of course).

Somehow I have trouble, the options don't seem to work like I expect them to. First of all, I'm a bit confused on how large objects should work: they are uploaded as MU in 5M parts, but downloaded in a single GET, so the limit of object size which determines whether object will be cached or not depends whether write-through cache from PUT operation or cache from another GET operation, is that right? This seems a bit strange if that's how it works. Or (in my case - objects rarely exceed 20M) - it's chunk size (5M), not total object that picks caching options?

Let's review caching options. These two are easy to set - I want gateway to work in normal mode, have 1 GB RAM object cache, no disk cache:
```
cache.http_cache = false
cache.cache_ram_capacity = 1073741824
cache.cache_disc_capacity = 0
```
This one is harder:
```
## When the length of the object exceeds this value, store the object on disk
### cache.cache_disc_threshold_len = 1048576
```
Is it talking about full object size or chunk size? Also, how does it work when having disk cache disabled - are objects larger than this threshold never cached if disk cache is disabled, or this option is silently ignored then? Confusing!

This option seems clear at first...
```
## Cache Max Content Length in bytes (http-cache)
cache.cache_max_content_len = 1048576
```
but then again, it mentions "(http-cache)", probably meaning that it only works when gateway is in http-cache mode, which is not the case. So looks like this option should not affect anything?

Experiment time. I assume that since this is object cache mode, and each object that goes through PUT never exceeds 5M (size of single chunk),  setting `cache.cache_disc_threshold_len = 6291456` would allow me to cache everything, at very least through write-through cache. Additionally, I set `cache.cache_max_content_len = 30728640`. But it fails. RAM usage for gateway increases very slowly, being stuck at 400 MB even through I've uploaded 10's GB through it already. And not in just large objects. It took days to climb to 600 MB, and about week later it's at 800 MB. I've uploaded 100's GBs, it cache should've been at max setting (1GB) in any case by now, even if doesn't store large objects at all, but it just doesn't grow to that value. From "status" output:
```
 [cache-related]
-------------------------------+------------------
       http cache [true/false] | false
       number of cache_workers | 16
                  cache expire | 300
         cache max content len | 30728640
            ram cache capacity | 1073741824
        disk cache capacity    | 0
        disk cache threshold   | 6291456
        disk cache data dir    | ./cache/data
        disk cache journal dir | ./cache/journal
-------------------------------+------------------
 [large object related]
-------------------------------+------------------
      max number of chunk objs | 1000
           chunk object length | 5242880
             max object length | 5242880000
     reading  chunk obj length | 5242880
     threshold of chunk length | 5767168
                               | 
      [PUT] worker pool size   | 64
      [PUT] worker buffer size | 32
```
From top:
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                     
22357 leofs     20   0 7657108 829964   3768 S   7,3  1,7 906:35.54 /usr/local/leofs/current/leo_gateway/erts-8.3.5.3/bin/beam+ 
```

Another gateway is configured with 60 MB for "disk cache threshold":  `cache.cache_disc_threshold_len = 62914560`. That one does manage to climb to 1GB in RAM usage. However, it took a long time for that. It took 10's GB of data uploaded through it for that to happen. So it doesn't try to store everything uploaded through it. Why?

In other words, I think there are 3 problems (last two might be related): options need better description, cache refuses to reach 1 GB unless cache_disc_threshold_len is greatly increased (even after uploading tons of small objects, far exceeding 1GB through that gateway), and cache filling seemingly far slower than it's supposed to.

**Comment by mocchira** [comment]
@vstax Thanks for filing this issue.


> Somehow I have trouble, the options don't seem to work like I expect them to. First of all, I'm a bit confused on how large objects should work: they are uploaded as MU in 5M parts, but downloaded in a single GET, so the limit of object size which determines whether object will be cached or not depends whether write-through cache from PUT operation or cache from another GET operation, is that right? This seems a bit strange if that's how it works. Or (in my case - objects rarely exceed 20M) - it's chunk size (5M), not total object that picks caching options?

First of all, now the cache is only available to small objects when cache_disc_capacity = 0. (this should be documented)
(might be off topic) For the record, when cache_disk_capacity > 0, larger objects can be cached as a file on the local file system. When those objects hit on cache, leofs return objects using sendfile to boost the performance.

The cache timing is right (write-through from PUT or cache from another GET).

> Is it talking about full object size or chunk size? Also, how does it work when having disk cache disabled - are objects larger than this threshold never cached if disk cache is disabled, or this option is silently ignored then? Confusing!

The full object size and objects which size are larger than cache_disc_threshold_len never be cached if cache_disc_capacity = 0. (documents should be refined to avoid confusing)

> but then again, it mentions "(http-cache)", probably meaning that it only works when gateway is in http-cache mode, which is not the case. So looks like this option should not affect anything?

Right, should not affect anything in your cases.

> Experiment time. I assume that since this is object cache mode, and each object that goes through PUT never exceeds 5M (size of single chunk), setting cache.cache_disc_threshold_len = 6291456 would allow me to cache everything, at very least through write-through cache. Additionally, I set cache.cache_max_content_len = 30728640. But it fails. RAM usage for gateway increases very slowly, being stuck at 400 MB even through I've uploaded 10's GB through it already. And not in just large objects. It took days to climb to 600 MB, and about week later it's at 800 MB. I've uploaded 100's GBs, it cache should've been at max setting (1GB) in any case by now, even if doesn't store large objects at all, but it just doesn't grow to that value. From "status" output:

This sounds problematic so I'll vet in further after 1.3.8 released (anyway which will be released today).

**Comment by mocchira** [comment]
@vstax Not directly related to this issue though let me ask one question here. After 1.3.8 released, we will get to work in 1.4 dev-cycle and at first we are going to check all issues opened and re-assign milestone on each of them based on our priorities so please let us know which issues should be fixed in 1.4. we will take your opinion into account as much as we can.

**Comment by vstax** [comment]
@mocchira 
> (might be off topic) For the record, when cache_disk_capacity > 0, larger objects can be cached as a file on the local file system. When those objects hit on cache, leofs return objects using sendfile to boost the performance.

I see. Unfortunately, it's a bad idea here as servers where I've launched gateways are kind of busy IO-wise (they have free memory and CPU though), so using disk cache would slow both gateway and other applications down. I have plenty of fast disk and memory cache set up at squid instead. The setup is like this:
```
storages <-> gateways <- Link between DC ->  (another DC) squid <-> applications (2)
                          \--> applications (1)
```
where applications (1) upload all the data and (sometimes) download it, and applications (2) - running in another datacenter building - download all freshly uploaded data a few times (once per application but there are 5-7 different types), and at times need to go through older data as well (when it happens, they download it multiple times like that).

So this setup is generally fine on caching, I only wanted to utilize some write-through cache from PUT operations on gateway because all information uploaded is requested almost right away.

> The full object size and objects which size are larger than cache_disc_threshold_len never be cached

Hmm but will it really work as write-through cache for large objects then? For example, let's say I have two gateways, G1 and G2 and disk cache is activated. Multipart upload (in few threads and connections) connects through LB and puts first part of object O - O_1 through G1 and second part - O_2 through G2. This means that part O_1 will be in cache of G1 and part O_2 will be in cache of G2 (plus header for O will be in one of the caches, depending on which gateway completed the upload, let's say it was G2). Now I request object O through G1. It gets header for O from storage (and puts it to cache), gets O_1 from cache, gets O_2 from storage (and puts it to cache). For next request of objects O, it will just verify cache against storage node and return object from cache if it's unchanged.. is that right? The reason I'm asking is that it feels a bit strange to have such (per-part) cache logic and that directive to look at full object size?

> This sounds problematic so I'll vet in further after 1.3.8 released (anyway which will be released today).

OK, thank you. Btw, if there is time-based object expiration (as per `cache expire | 300` option) this problem could be explained, but that option should only be active for http-cache, as I understand. Not to mention that increasing  `cache.cache_disc_threshold_len` from 6 MB to 60 MB shouldn't affect anything with disk cache disabled (since there are no single-part objects larger than 5M), but it does change caching behaviour for me.

> Not directly related to this issue though let me ask one question here. After 1.3.8 released, we will get to work in 1.4 dev-cycle and at first we are going to check all issues opened and re-assign milestone on each of them based on our priorities so please let us know which issues should be fixed in 1.4. we will take your opinion into account as much as we can.

Thank you; I appreciate you taking it into account! I posted a list with notices here: https://gist.github.com/vstax/444ec8fc53c3b0a0bbd515146c1031cb (so there won't be meaningless mentions of this issue). There are a few other minor issues as well that I'm interested in but seem to be all are assigned to 1.4 already.

**Comment by mocchira** [comment]
@vstax

> Hmm but will it really work as write-through cache for large objects then? For example, let's say I have two gateways, G1 and G2 and disk cache is activated. Multipart upload (in few threads and connections) connects through LB and puts first part of object O - O_1 through G1 and second part - O_2 through G2. This means that part O_1 will be in cache of G1 and part O_2 will be in cache of G2 (plus header for O will be in one of the caches, depending on which gateway completed the upload, let's say it was G2). Now I request object O through G1. It gets header for O from storage (and puts it to cache), gets O_1 from cache, gets O_2 from storage (and puts it to cache). For next request of objects O, it will just verify cache against storage node and return object from cache if it's unchanged.. is that right? The reason I'm asking is that it feels a bit strange to have such (per-part) cache logic and that directive to look at full object size?

Sorry, my words weren't enough and also the spec is too complicated!
Now there are two different cache mechanisms for each (small and large objects).
- Caching for small objects
  - Cache happens in both cases PUT(write through) and GET
  - Stored to memory if the size of an object < cache.cache_disc_threshold_len
  - Stored to DISK if the size of an object >= cache.cache_disc_threshold_len
- Caching for large objects
  - Cache happens ONLY in case GET
  - Stored to always DISK ONLY if disk cache is enabled
  - the whole part object is written as a file on the host file system during processing a GET request

> OK, thank you. Btw, if there is time-based object expiration (as per cache expire | 300 option) this problem could be explained, but that option should only be active for http-cache, as I understand. Not to mention that increasing cache.cache_disc_threshold_len from 6 MB to 60 MB shouldn't affect anything with disk cache disabled (since there are no single-part objects larger than 5M), but it does change caching behaviour for me.

Yes time-based expiration is active only for http-cache so there must be another culprit. (WIP)

> Thank you; I appreciate you taking it into account! I posted a list with notices here: https://gist.github.com/vstax/444ec8fc53c3b0a0bbd515146c1031cb (so there won't be meaningless mentions of this issue). There are a few other minor issues as well that I'm interested in but seem to be all are assigned to 1.4 already.

Thanks for sharing. We will assign as much those issues into 1.4 milestone as possible. anyway just in case, let us know  a few other minor issues you have interests because we might re-assign the milestone which already set to 1.4 into something other (1.5 or later) for the shipping schedule.

**Comment by mocchira** [comment]
Postpone the investigation part of this issue to 1.4.1 and fixing cache documents will be filed as another issue and set its milestone to 1.4.0.

**Comment by mocchira** [comment]
filed https://github.com/leo-project/leofs/issues/997 for fixing documents.
